documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md

```diff
@@ -6,283 +6,268 @@ Proposal design for a platform invoke (P/Invoke) mechanism for Silk.NET 3.0.
+public class MyStringLibrary : IMyStringLibrary
 {
-    [UseDynamicLibrary("glfw3.dll")]
-    public partial nint glfwGetProcAddress(byte* str);
+    public static class DllImport : IMyStringLibrary.Static
+    {
+        [DllImportAttribute("mystringlib")]
+        public static extern byte* ToLower(byte* str);
+        [DllImportAttribute("mystringlib")]
+        public static extern void FreeResult(byte* str);
+    }
+    public class StaticWrapper<T> : IMyStringLibrary where T: IMyStringLibrary.Static
+    {
+        public StaticWrapper();
+        public byte* ToLower(byte* str) => T.ToLower(str);
+        public void FreeResult(byte* str) => T.FreeResult(str);
+    }
+    public class ThreadLocal : IMyStringLibrary.Static
+    {
+        private static ThreadLocal<IMyStringLibrary> _current = new();
+        public static void MakeCurrent(IMyStringLibrary current) => _current.Value = current;
+        public static byte* ToLower(byte* str) => _current.Value.ToLower(str);
+        public static void FreeResult(byte* str) => _current.Value.FreeResult(str);
+        public static Ptr<byte> ToLower(Ptr<byte> str) => _current.Value.ToLower(str);
+        public static void FreeResult(Ptr<byte> str) => _current.Value.FreeResult(str);
+    }
+
+    // Non-Static Interface
+    private Func<string, nint> _getProcAddress;
+    byte* IMyStringLibrary.ToLower(byte* str)
+    {
+        var ptr = _getProcAddress("ToLower");
+        if (ptr is 0) throw new("some symbol loading exception...");
+        return ((delegate* unmanaged<byte*, byte*>)ptr)(str);
+    }
+    
+    void IMyStringLibrary.FreeResult(byte* str)
+    {
+        var ptr = _getProcAddress("FreeResult");
+        if (ptr is 0) throw new("some symbol loading exception...");
+        return ((delegate* unmanaged<byte*, void>)ptr)(str);
+    }
 
-    // shim to convert the string, which SilkTouch needs to use, to a byte pointer - THIS IS NOT A MODEL EXAMPLE!
-    private nint GetProcAddressShim(string str) => glfwGetProcAddress((byte*) Marshal.StringToHGlobalAnsi(str));
+    public static IMyStringLibrary Create() => new StaticWrapper<DllImport>();
+    public static IMyStringLibrary Create(Func<string, nint> getProcAddress) => new MyStringLibrary { _getProcAddress = getProcAddress };
 
-    public partial void glBegin(uint mode);
+    // Static Interface
+    public static byte* ToLower(byte* str) => DllImport.ToLower(str);
+    public static Ptr<byte> ToLower(Ptr<byte> str) => DllImport.ToLower(str);
+    public static void FreeResult(byte* str) => DllImport.FreeResult(str);
+    public static void FreeResult(Ptr<byte> str) => DllImport.FreeResult(str);
 }
 ```
 
-Procedure Address Methods are method groups (or an otherwise callable expression) within the scope of the method that **MUST** return a `void*`, `nint`, or `IntPtr` when invoked. This is the actual address in memory of the function being invoked.
+**INFORMATIVE:** `getProcAddress` delegates replace "native contexts".
```

Comment:
Maxine has expressed interest in Discord in not doing this, and maintaining an INativeContext interface in some form rather than forcing delegates to be used.